File: 07_Web_Scraping.R
## TIP and WARNING: We highly encourage you to save you web scraped results

## load packages
library(janitor)
library(tidyverse)
library(rvest)
library(htmlTable)
library(zoo)
library(kableExtra)

## web scrape
chap_7_file <- "./data/chap_7_all_r_file.csv"
if (!file.exists(chap_7_file)) {
    draft_r <- tibble()
    for (i in seq(from = 2000, to = 2022)) {
        url <- paste0(
            "https://www.pro-football-reference.com/years/",
            i,
            "/draft.htm"
        )
        web_data <-
            read_html(url) |>
            html_nodes(xpath = '//*[@id="drafts"]') |>
            html_table()
        web_df <-
            web_data[[1]]
        web_df_clean <-
            web_df |>
            janitor::row_to_names(row_number = 1) |>
            janitor::clean_names(case = "none") |>
            mutate(Season = i) |> # add seasons
            filter(Tm != "Tm") # Remove any extra column headers
        draft_r <- bind_rows(draft_r, web_df_clean)
    }
    write_csv(draft_r, chap_7_file)
    draft_r <- read_csv(chap_7_file)
} else {
    draft_r <- read_csv(chap_7_file)
}

## rename teams
# the chargers moved to Los Angeles from San Diego
# the Raiders moved from Oakland to Las Vegas
# the Rams moved from St. Louis to Los Angeles
draft_r <-
    draft_r |>
    mutate(
        Tm = case_when(
            Tm == "SDG" ~ "LAC",
            Tm == "OAK" ~ "LVR",
            Tm == "STL" ~ "LAR",
            TRUE ~ Tm
        ),
        DrAV = ifelse(is.na(DrAV), 0, DrAV)
    )

draft_r_use <-
    draft_r |>
    select(Season, Pick, Tm, Player, Pos, wAV, DrAV)
print(draft_r_use)

## plot DrAV
draft_r_use_pre2019 <- draft_r_use |>
    mutate(
        DrAV = as.numeric(DrAV),
        wAV = as.numeric(wAV),
        Pick = as.integer(Pick)
    ) |>
    filter(Season <= 2019)

ggplot(draft_r_use_pre2019, aes(Pick, DrAV)) +
    geom_point(alpha = 0.2) +
    stat_smooth() +
    theme_bw()

## DrAV each position
draft_chart_r <-
    draft_r_use_pre2019 |>
    group_by(Pick) |>
    summarize(mean_DrAV = mean(DrAV, na.rm = TRUE)) |>
    mutate(
        mean_DrAV = ifelse(is.na(mean_DrAV), 0, mean_DrAV)
    ) |>
    mutate(
        roll_DrAV = rollapply(mean_DrAV,
            width = 13, FUN = mean,
            na.rm = TRUE, fill = "extend", partial = TRUE
        )
    )

ggplot(draft_chart_r, aes(Pick, roll_DrAV)) +
    geom_point() +
    geom_smooth() +
    theme_bw() +
    ylab("Rolling average (\u00B1 6) DrAV") +
    xlab("Draft pick")

DrAV_pick_fit_r <-
    draft_chart_r |>
    lm(formula = log(roll_DrAV + 1) ~ Pick)
summary(DrAV_pick_fit_r)

draft_chart_r <-
    draft_chart_r |>
    mutate(
        fitted_DrAV = pmax(0, exp(predict(DrAV_pick_fit_r)) - 1)
    )
draft_chart_r |>
    head()

## Jets/Colts 2018 trade evaluated
future_pick <-
    tibble(
        Pick = "Future 2nd round",
        Value = "14.8 (discounted at rate of 25%)"
    )

team <- tibble("Receiving team" = c("Jets", rep("Colts", 4)))

tbl_1 <-
    draft_chart_r |>
    filter(Pick %in% c(3, 6, 37, 49)) |>
    select(Pick, fitted_DrAV) |>
    rename(Value = fitted_DrAV) |>
    mutate(Pick = as.character(Pick), Value = as.character(round(Value, 1))) |>
    bind_rows(future_pick)

team |>
    bind_cols(tbl_1) |>
    kbl(format = "pipe") |>
    kable_styling()

future_pick <-
    tibble(
        Pick = "Future 2nd round",
        Value = "14.8 (discounted at rate of 25)"
    )

results_trade <-
    tibble(
        Team = c("Jets", rep("Colts", 5)),
        Pick = c(
            3, 6, 37,
            "49-traded for 52",
            "49-traded for 169",
            "52 in 2019"
        ),
        Player = c(
            "Sam Darnold", "Quenton Nelson", "Braden Smith",
            "Kemoko Turay", "Jordan Wilkins", "Rock Ya-Sin"
        ),
        "DrAV" = c(25, 55, 32, 5, 8, 11)
    )

results_trade |>
    kbl(format = "pipe") |>
    kable_styling()


## Are some teams better at drafting?
## R draft
draft_r_use_pre2019 <-
    draft_r_use_pre2019 |>
    left_join(draft_chart_r |>
        select(Pick, fitted_DrAV), by = "Pick")

draft_r_use_pre2019 |>
    group_by(Tm) |>
    summarize(
        total_picks = n(),
        DrAV_OE = mean(DrAV - fitted_DrAV, na.rm = TRUE),
        DrAV_sigma = sd(DrAV - fitted_DrAV, na.rm = TRUE)
    ) |>
    arrange(-DrAV_OE) |>
    print(n = Inf)

draft_r_use_pre2019 |>
    group_by(Tm) |>
    summarize(
        total_picks = n(),
        DrAV_OE = mean(DrAV - fitted_DrAV, na.rm = TRUE),
        DrAV_sigma = sd(DrAV - fitted_DrAV, na.rm = TRUE)
    ) |>
    mutate(
        se = DrAV_sigma / sqrt(total_picks),
        lower_bound = DrAV_OE - 1.96 * se,
        upper_bound = DrAV_OE + 1.96 * se
    ) |>
    arrange(-DrAV_OE) |>
    print(n = Inf)

File: 07_Web_Scraping.py
## TIP and WARNING: We highly encourage you to save you web scraped results

## Load packages
import pandas as pd
import seaborn as sns
import os.path
import matplotlib.pyplot as plt
import statsmodels.formula.api as smf
import numpy as np
import os

## download data
## be careful to hit webpage too many time and get locked out~
url = "https://www.pro-football-reference.com/years/2022/draft.htm"

chap_7_2022_py_file = "chap_7_2022_py_file.csv"

if os.path.isfile(chap_7_2022_py_file):
    draft_py = pd.read_csv(chap_7_2022_py_file)
else:
    draft_py = pd.read_html(url, header=1)[0]
    draft_py.to_csv(chap_7_2022_py_file)


draft_py.loc[draft_py["DrAV"].isnull(), "DrAV"] = 0

print(draft_py)

draft_py[["Rnd", "Pick", "Tm", "Sk", "College/Univ", "Unnamed: 28"]]

## loop over multiple years

chap_7_all_py_file = "chap_7_all_py_file.csv"

if os.path.isfile(chap_7_all_py_file):
    draft_py = pd.read_csv(chap_7_all_py_file)
else:
    draft_py = pd.DataFrame()
    for i in range(2000, 2022 + 1):
        url = "https://www.pro-football-reference.com/years/" + str(i) + "/draft.htm"
        web_data = pd.read_html(url, header=1)[0]
        web_data["Season"] = i
        web_data = web_data.query('Tm != "Tm"')
        draft_py = pd.concat([draft_py, web_data])
    draft_py.to_csv(chap_7_all_py_file)

draft_py.reset_index(drop=True, inplace=True)

## rename teams
# the Chargers moved to Los Angeles from San Diego
# the Raiders moved from Oakland to Las Vegas
# the Rams moved from St. Louis to Los Angeles

conditions = [
    (draft_py.Tm == "SDG"),
    (draft_py.Tm == "OAK"),
    (draft_py.Tm == "STL"),
]
choices = ["LAC", "LVR", "LAR"]
draft_py["Tm"] = np.select(conditions, choices, default=draft_py.Tm)

## replace missing DrAV with 0
draft_py.loc[draft_py["DrAV"].isnull(), "DrAV"] = 0
draft_py.to_csv("data_py.csv", index=False)

print(draft_py.head())

print(draft_py.columns)

## only look at select columns of interest
draft_py_use = draft_py[["Season", "Pick", "Tm", "Player", "Pos", "wAV", "DrAV"]]

print(draft_py_use)

## Plot for DrAV
# Change theme for chapter
sns.set_theme(style="whitegrid", palette="colorblind")

draft_py_use_pre2019 = draft_py_use.query("Season <= 2019")

## format columns as numeric or integers
draft_py_use_pre2019 = draft_py_use_pre2019.astype({"Pick": int, "DrAV": float})

sns.regplot(
    data=draft_py_use_pre2019,
    x="Pick",
    y="DrAV",
    line_kws={"color": "red"},
    scatter_kws={"alpha": 0.2},
)
plt.show()

## DrAV for each pick position
draft_chart_py = draft_py_use_pre2019.groupby(["Pick"]).agg({"DrAV": ["mean"]})
draft_chart_py.columns = list(map("_".join, draft_chart_py.columns))
draft_chart_py.loc[draft_chart_py.DrAV_mean.isnull()] = 0
draft_chart_py["roll_DrAV"] = (
    draft_chart_py["DrAV_mean"].rolling(window=13, min_periods=1, center=True).mean()
)

sns.scatterplot(draft_chart_py, x="Pick", y="roll_DrAV")
plt.show()

## regression
draft_chart_py.reset_index(inplace=True)
draft_chart_py["roll_DrAV_log"] = np.log(draft_chart_py["roll_DrAV"] + 1)
DrAV_pick_fit_py = smf.ols(formula="roll_DrAV_log ~ Pick", data=draft_chart_py).fit()
print(DrAV_pick_fit_py.summary())

draft_chart_py["fitted_DrAV"] = np.exp(DrAV_pick_fit_py.predict()) - 1
draft_chart_py.head()

## are some teams better at drafting?
draft_py_use_pre2019 = draft_py_use_pre2019.merge(
    draft_chart_py[["Pick", "fitted_DrAV"]], on="Pick"
)
draft_py_use_pre2019["OE"] = (
    draft_py_use_pre2019["DrAV"] - draft_py_use_pre2019["fitted_DrAV"]
)
draft_py_use_pre2019.groupby("Tm").agg(
    {"OE": ["count", "mean", "std"]}
).reset_index().sort_values([("OE", "mean")], ascending=False)

draft_py_use_pre2019 = draft_py_use_pre2019.merge(
    draft_chart_py[["Pick", "fitted_DrAV"]], on="Pick"
)

draft_py_use_pre2019_tm = (
    draft_py_use_pre2019.groupby("Tm")
    .agg({"OE": ["count", "mean", "std"]})
    .reset_index()
    .sort_values([("OE", "mean")], ascending=False)
)

draft_py_use_pre2019_tm.columns = list(map("_".join, draft_py_use_pre2019_tm.columns))
draft_py_use_pre2019_tm.reset_index(inplace=True)

draft_py_use_pre2019_tm["se"] = draft_py_use_pre2019_tm["OE_std"] / np.sqrt(
    draft_py_use_pre2019_tm["OE_count"]
)

draft_py_use_pre2019_tm["lower_bound"] = (
    draft_py_use_pre2019_tm["OE_mean"] - 1.96 * draft_py_use_pre2019_tm["se"]
)

draft_py_use_pre2019_tm["upper_bound"] = (
    draft_py_use_pre2019_tm["OE_mean"] + 1.96 * draft_py_use_pre2019_tm["se"]
)

print(draft_py_use_pre2019_tm)
